numpy/f2py/cfuncs.py

```diff
@@ -621,71 +630,118 @@
 }"""
 needs['try_pyarr_from_string'] = ['STRINGCOPYN', 'PRINTPYOBJERR', 'string']
 cfuncs['try_pyarr_from_string'] = """\
-static int try_pyarr_from_string(PyObject *obj,const string str) {
+/*
+  try_pyarr_from_string copies str[:len(obj)] to a mutable Python
+  string-like obj.
+
+  if the specified len==-1, str must be null-terminated.
+*/
+static int try_pyarr_from_string(PyObject *obj,
+                                 const string str, const int len) {
+#ifdef DEBUGCFUNCS
+fprintf(stderr,\"try_pyarr_from_string(str='%s',
+               len=%d, obj=%p)\\n\",(char*)str,len, obj);
+#endif
     PyArrayObject *arr = NULL;
-    if (PyArray_Check(obj) && (!((arr = (PyArrayObject *)obj) == NULL)))
-        { STRINGCOPYN(PyArray_DATA(arr),str,PyArray_NBYTES(arr)); }
-    return 1;
+    if (PyArray_Check(obj) && (!((arr = (PyArrayObject *)obj) == NULL))) {
+        string buf = PyArray_DATA(arr);
+        int n = len;
+        if (n == -1) {
+            /* Assuming null-terminated str. */
+            n = strlen(str);
+        }
+        if (n > PyArray_NBYTES(arr)) {
+            n = PyArray_NBYTES(arr);
+            if (n == 1 && buf[0] == '\0') {
+                /* handle the numpy.array(b'') case */
```

Comment:
I don't understand why this needs a special case

Reply:
This is related to the following numpy behavior:
```
>>> len(np.array(b'').tobytes())
1
>>> len(np.array(b'A').tobytes())
1
```
that is, a zero-length `bytes` object consumes 1 byte in numpy array (and it holds zero-terminator) while passing the corresponding buffer to Fortran function ought to have 0-byte length. Or stated differently, the output of `PyArray_NBYTES(np.array(b''))` is wrong in the context of Fortran character arrays that are zero-terminator ignorant.

Reply:
> hat is, a zero-length bytes object consumes 1 byte in numpy array

I disagree with this, all you're seeing is that the default dtype for byte arrays is `B{min(1,len(b))}`. Other ways of constructing an array do not add this trailing null byte:
```
>>> b0 = np.empty(1, [('x', 'S0')])['x']
>>> b0
array([b''], dtype='|S0')
>>> len(b0.tobytes())
0
```

Reply:
I agree that my statement was inaccurate but I doubt that users will figure out to use the proposed construction for handling zero-length byte arrays. Users would most likely use `np.array(b'')`. This changeset ensures that users will have the expected behavior on the Fortran side despite, I would say, the NumPy weird behavior.

In fact, it appears to be impossible (if it is not true then read: I have not found out how) to create `array(b'', dtype='|S0')`. Here are my attempts:
```
>>> np.array(b'', dtype='|S0')
array(b'', dtype='|S1')
>>> np.array(b'', dtype=np.dtype('S0'))
array(b'', dtype='|S1')
>>> dtype = np.empty(1, [('x', 'S0')])['x'].dtype
>>> dtype
dtype('S')
>>> np.array(b'', dtype=dtype)
array(b'', dtype='|S1')
```

Reply:
You can create such an array with `b0 = np.empty((), [('x', 'S0')])['x']`.

> Users would most likely use `np.array(b'')`.

Wouldn't they just use `b''` without the array at all? I'm not sure I understand where this function is actually exposed to the user yet.

Reply:
No, one must use ndarray for `intent(inout)` arguments because `bytes` objects are immutable.

Reply:
Can you add a test with the malicious b0 I show above, just to show that nothing goes wrong?

Reply:
Good idea, done.

Reply:
In case your interested, an easier way to get hold of such an array is `np.ndarray((), np.bytes_, buffer=b'')`

Reply:
The special case is now unhandled.